wire-ios-data-model/Source/ManagedObjectContext/Migration/2.106-2.107/CleanupModels2_107PreAction.swift

```diff
@@ -17,29 +17,44 @@
 //
 
 import Foundation
-import WireSystem
 
-class DuplicateClientsMigrationPolicy: NSEntityMigrationPolicy {
+/// Removes UserClient duplicates and invalid ParticipantRoles
+class CleanupModels2_107PreAction: CoreDataMigrationAction {
 
     private enum Keys: String {
+        case conversation
         case needsToBeUpdatedFromBackend
     }
 
-    private let zmLog = ZMSLog(tag: "core-data")
+    override func execute(in context: NSManagedObjectContext) throws {
 
-    override func begin(_ mapping: NSEntityMapping, with manager: NSMigrationManager) throws {
-        zmLog.safePublic("beginning duplicate clients migration", level: .info)
-        WireLogger.localStorage.info("beginning duplicate clients migration")
+        try removeInvalidParticipantRoles(in: context)
+        try removeUserClientDuplicates(in: context)
+    }
+
+    private func removeInvalidParticipantRoles(in context: NSManagedObjectContext) throws {
+        let entityName = ParticipantRole.entityName()
+        // we don't want to load the ParticipantRole model here just plain NSManagedObject in case of changes in future
+        let request = NSFetchRequest<NSManagedObject>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == nil", Keys.conversation.rawValue)
 
-        let context = manager.sourceContext
+        let result = try context.fetch(request)
+        for object in result {
+            WireLogger.localStorage.warn("remove zombie object 'ParticipantRole' without conversation", attributes: .safePublic)
+            context.delete(object)
```

Comment:
After such action should we perform a `context.save`? Or does it happen later somewhere?

Reply:
context.save is called in the perform method of the super class